src/Microsoft.VisualStudio.ProjectSystem.Managed/Threading/Tasks/CancellationSeries.cs

```diff
@@ -0,0 +1,112 @@
+
+using System;
+using System.Threading;
+
+namespace Microsoft.VisualStudio.Threading.Tasks
+{
+    /// <summary>
+    /// Produces a series of <see cref="CancellationToken"/> objects such that requesting a new token
+    /// causes the previously issued token to be cancelled.
+    /// </summary>
+    /// <remarks>
+    /// <para>Consuming code is responsible for managing overlapping asynchronous operations.</para>
+    /// <para>This class has a lock-free implementation to minimise latency and contention.</para>
+    /// </remarks>
+    public sealed class CancellationSeries : IDisposable
+    {
+        private CancellationTokenSource _cts = new CancellationTokenSource();
+
+        private readonly CancellationToken[] _superToken;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="CancellationSeries"/>.
+        /// </summary>
+        /// <param name="token">An optional cancellation token that, when cancelled, cancels the last
+        /// issued token and causes any subsequent tokens to be issued in a cancelled state.</param>
+        public CancellationSeries(CancellationToken token = default)
+        {
+            _superToken = token == default 
+                ? null 
+                : new[] { token };
+        }
+
+        /// <summary>
+        /// Gets the next <see cref="CancellationToken"/> in the series, ensuring the last issued
+        /// token (if any) is cancelled before this method returns.
+        /// </summary>
+        /// <param name="token">An optional cancellation token that, when cancelled, cancels the
+        /// returned token.</param>
+        /// <returns>
+        /// A cancellation token that will be cancelled when either:
+        /// <list type="bullet">
+        /// <item><see cref="GetToken"/> is called again</item>
+        /// <item>The token passed to this method (if any) is cancelled</item>
+        /// <item>The token passed to the constructor (if any) is cancelled</item>
+        /// <item><see cref="Dispose"/> is called</item>
+        /// </list>
+        /// </returns>
+        /// <exception cref="ObjectDisposedException">This object has been disposed.</exception>
+        public CancellationToken GetToken(CancellationToken token = default)
+        {
+            // Create the next CTS. There are four scenarios, depending upon whether
+            // a super token was given to the constructor, and whether a token was
+            // passed to this method.
+
+            CancellationTokenSource nextSource = _superToken == null
+                ? token == default
+                    ? new CancellationTokenSource()
+                    : CancellationTokenSource.CreateLinkedTokenSource(token)
+                : token == default
+                    ? CancellationTokenSource.CreateLinkedTokenSource(_superToken)
+                    : CancellationTokenSource.CreateLinkedTokenSource(token, _superToken[0]);
+
+            CancellationToken nextToken = nextSource.Token;
```

Comment:
Why store this into a local, rather than just `return nextSource.Token` at the end?

Reply:
No reason. Fixed.